scene/resources/mesh_data_tool.cpp

```diff
@@ -446,6 +515,69 @@ void MeshDataTool::set_vertex_meta(int p_idx, const Variant &p_meta) {
 	vertices.write[p_idx].meta = p_meta;
 }
 
+Variant MeshDataTool::get_vertex_custom(int p_idx, int p_cidx) const {
+	ERR_FAIL_INDEX_V(p_idx, vertices.size(), Variant());
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, Variant());
+	uint32_t type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+	if (type > Mesh::ARRAY_CUSTOM_RGBA_HALF) {
+		return vertices[p_idx].custom[p_cidx].fd;
+	} else {
+		Vector<uint8_t> ret;
+		ret.resize(8);
+		memcpy(ret.ptrw(), vertices[p_idx].custom[p_cidx].ud, 8);
+		return ret;
+	}
+}
+
+Error MeshDataTool::set_vertex_custom(int p_idx, int p_cidx, const Color &p_customf, const PackedByteArray &p_customu) {
+	ERR_FAIL_INDEX_V(p_idx, vertices.size(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, ERR_INVALID_PARAMETER);
+
+	if (format & (Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx)) {
+		uint32_t type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+		if (type > Mesh::ARRAY_CUSTOM_RGBA_HALF) {
+			vertices.write[p_idx].custom[p_cidx].fd = p_customf;
```

Comment:
I'm scared that this is duplicate code with the else case below...

this logic is quite complex.

Reply:
it is not duplicate code.
line 536 will check to see if we already have custom data for our custom index.
line 537 extracts the format of that custom index. (it might be a zero number which indicates `ARRAY_CUSTOM_RGBA8_UNORM`.
line 538 will try to see if it's a floating point type or not.